### PR TITLE
Add json output

### PR DESCRIPTION
**Pull request by ecraven**
Add json output (compatible with what koreader sdcv has), but without a library.

### CODE REVIEW

**Comment by ecraven** [comment]
see #6

**Comment by ecraven** [comment]
Everything should work fine, only the order of the dictionaries in json is different, so I need to adjust the test for that. Please review the PR, I'll fix the test, but I'll only be back at a computer next week :-/

File: src/libwrapper.cpp [hunk 1]
```diff
@@ -257,18 +257,30 @@ void Library::print_search_result(FILE *out, const TSearchResult & res)
...
-            colorize_output_ ? NAME_OF_DICT_VISFMT : "",
-            utf8_output_ ? res.bookname.c_str() : loc_bookname.c_str(),
-            colorize_output_ ? ESC_END : "",
-            colorize_output_ ? SEARCH_TERM_VISFMT : "",
-            utf8_output_ ? res.def.c_str() : loc_def.c_str(),
-            colorize_output_ ? ESC_END : "",
-            utf8_output_ ? res.exp.c_str() : loc_exp.c_str());
+        if(use_json) {
+          if(!*first_result) {
+            fprintf(out, ",");
```
Comment (Dushistov) [inline src/libwrapper.cpp]: `fputc` will be little faster in this case

File: src/libwrapper.cpp [hunk 2]
```diff
@@ -248,7 +248,7 @@ void Library::LookupData(const std::string &str, TSearchResultList& res_list)
 		}
 }
 
-void Library::print_search_result(FILE *out, const TSearchResult & res)
+void Library::print_search_result(FILE *out, const TSearchResult & res, bool use_json, bool *first_result)
```
Comment (Dushistov) [inline src/libwrapper.cpp]: see comment about `use_json` bellow

File: src/libwrapper.cpp [hunk 3]
```diff
@@ -296,7 +308,7 @@ namespace {
     };
 }
 
-bool Library::process_phrase(const char *loc_str, IReadLine &io, bool force)
+bool Library::process_phrase(const char *loc_str, IReadLine &io, bool force, bool use_json)
```
Comment (Dushistov) [inline src/libwrapper.cpp]: see comment about `use_json` bellow

File: src/libwrapper.cpp [hunk 4]
```diff
@@ -346,6 +358,11 @@ bool Library::process_phrase(const char *loc_str, IReadLine &io, bool force)
 		/*nothing*/;
 	}
 
+	sdcv_pager pager(force);
+	bool first_result = true;
+	if(use_json) {
+	  fprintf(pager.get_stream(), "[");
```
Comment (Dushistov) [inline src/libwrapper.cpp]: `fputc` will be little faster

File: src/libwrapper.cpp [hunk 5]
```diff
@@ -416,5 +434,8 @@ bool Library::process_phrase(const char *loc_str, IReadLine &io, bool force)
 		printf(_("Nothing similar to %s, sorry :(\n"), utf8_output_ ? get_impl(str) : loc_str.c_str());
 	}
 
+        if(use_json) {
+          fprintf(pager.get_stream(), "]\n");
```
Comment (Dushistov) [inline src/libwrapper.cpp]: `fputs` would be little faster

File: src/libwrapper.hpp [hunk 6]
```diff
@@ -28,7 +28,7 @@ class Library : public Libs {
 	Library(bool uinput, bool uoutput, bool colorize_output):
 		utf8_input_(uinput), utf8_output_(uoutput), colorize_output_(colorize_output) {}
 
-	bool process_phrase(const char *loc_str, IReadLine &io, bool force = false);
+	bool process_phrase(const char *loc_str, IReadLine &io, bool force = false, bool use_json = false);
 private:
 	bool utf8_input_;
     bool utf8_output_;
```
Comment (Dushistov) [inline src/libwrapper.hpp]: why not make `use_json` class member of `Library`, like `utf8_input_` and `utf8_output_`?

File: src/sdcv.cpp [hunk 7]
```diff
@@ -220,17 +227,32 @@ int main(int argc, char *argv[]) try {
...
     printf(_("Dictionary's name   Word count\n"));
-    std::list<std::string> order_list, disable_list;
-    for_each_file(dicts_dir_list, ".ifo", order_list,
-                  disable_list, [](const std::string &filename, bool) -> void {
-                      DictInfo dict_info;
-                      if (dict_info.load_from_ifo_file(filename, false)) {
-                          const std::string bookname = utf8_to_locale_ign_err(dict_info.bookname);
-                          printf("%s    %d\n", bookname.c_str(), dict_info.wordcount);
+  else
+    printf("[");
```
Comment (Dushistov) [inline src/sdcv.cpp]: `fputc`

File: src/sdcv.cpp [hunk 8]
```diff
@@ -220,17 +227,32 @@ int main(int argc, char *argv[]) try {
...
                       }
-                  });
+                      printf("{\"name\": \"%s\", \"wordcount\": \"%d\"}", json_escape_string(bookname).c_str(), dict_info.wordcount);
+                    } else {
+                      printf("%s    %d\n", bookname.c_str(), dict_info.wordcount);
+                    }
+                  }
+                });
+  if(use_json)
+    printf("]\n");
```
Comment (Dushistov) [inline src/sdcv.cpp]: `fputs`

File: src/sdcv.cpp [hunk 9]
```diff
@@ -220,17 +227,32 @@ int main(int argc, char *argv[]) try {
...
+  for_each_file(dicts_dir_list, ".ifo", order_list,
+                disable_list, [use_json, &first_entry](const std::string &filename, bool) -> void {
+                  DictInfo dict_info;
+                  if (dict_info.load_from_ifo_file(filename, false)) {
+                    const std::string bookname = utf8_to_locale_ign_err(dict_info.bookname);
+                    if(use_json) {
+                      if(first_entry) {
+                        first_entry=false;
+                      } else {
+                        printf(","); // comma between entries
```
Comment (Dushistov) [inline src/sdcv.cpp]: `fputc`

File: src/libwrapper.cpp (hunk 2 above)
Comment (Dushistov) [inline src/libwrapper.cpp]: better use `bool &` to prevent possible `nullptr`

Comment (ecraven) [inline src/libwrapper.cpp]: done, I've moved this to Library::json_

Comment (ecraven) [inline src/libwrapper.cpp]: thanks, fixed.

File: src/libwrapper.cpp (hunk 1 above)
Comment (ecraven) [inline src/libwrapper.cpp]: done, I've used fputs and fputc everywhere ;) thanks for the pointer!

**Comment by ecraven** [comment]
OK, tests finish fine now, if possible please review again.
I've included 5 or so whitespace-only indentation changes, sorry for not putting them into their own commit.
json output should work fine now, the "Nothing similar to..." and "save to ..." messages are *not* shown if -j is active.

File: tests/t_json [hunk 10]
```diff
@@ -0,0 +1,38 @@
+#!/bin/sh
```
Comment (Dushistov) [inline tests/t_json]: All three test cases looks similar, can you move into sh/bash function/functions?
To simplify ading new test cases in the future?

Comment (Dushistov) [inline tests/t_json]: You can use `tests/t_synonyms` as example,
but in your case I suppose function should accept two arguments:
- `sdcv` argument (may be empty)
- expected json output

Comment (ecraven) [inline tests/t_json]: ok, should be fixed ;)

**Comment by ecraven** [comment]
just to note it somewhere, I've included jq in the travis file, and use it to normalize and sort all json output, so that it can be compared better.

**Pull request merged**